web/source/kmwembedded.ts

```diff
@@ -111,42 +111,42 @@ namespace com.keyman.osk {
    * Adjust the absolute height of each keyboard element after a rotation - modified for KMEI, 13/11/14
    *    
    **/      
-  VisualKeyboard.prototype.adjustHeights=function(this: VisualKeyboard): boolean {
+  VisualKeyboard.prototype.adjustHeights = function(this: VisualKeyboard): boolean {
     let keyman = com.keyman.singleton;
-    let osk = keyman.osk;
-    let _Box = osk._Box;
+    let oskManager = keyman.osk;
+    let _Box = oskManager._Box;
     let util = keyman.util;
     let device = util.device;
 
     if(!_Box || !this.kbdDiv || !this.kbdDiv.firstChild || !this.kbdDiv.firstChild.firstChild.childNodes) {
       return false;
     }
 
-    var layers=_Box.firstChild.firstChild.childNodes,
+    var layers=this.kbdDiv.firstChild.childNodes,
         nRows=layers[0].childNodes.length,
-        oskHeight=osk.getHeight(),
-        rowHeight=Math.floor(oskHeight/nRows),
+        oskHeight=oskManager.getHeight(),
+        bannerHeight=oskManager.banner.height,
+        rowHeight=Math.floor((oskHeight-bannerHeight)/(nRows == 0 ? 1 : nRows)),
         nLayer,nRow,rs,keys,nKeys,nKey,key,ks,j,pad=4,fs=1.0;
 
     if(device.OS == 'Android' && 'devicePixelRatio' in window) {
       rowHeight = rowHeight/window.devicePixelRatio;
     }
+    oskHeight=nRows*rowHeight+bannerHeight;
 
-    oskHeight=nRows*rowHeight;
-
-    var b: HTMLElement = osk._Box, bs=b.style;
+    var b: HTMLElement = _Box, bs=b.style;
     bs.height=bs.maxHeight=(oskHeight+3)+'px';
-    b = <HTMLElement> b.firstChild.firstChild;
+    b = <HTMLElement> b.childNodes.item(1).firstChild;
```

Comment:
OK, so since the banner bar is now the `firstChild` on mobile form factors, we need `item(1)` to reference the visual keyboard?

Reply:
That's right